Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -137,7 +137,7 @@ public class XPCClient {
         xpc_connection_resume(connection)
 
         switch endpoint.kind {
-        case .xpcServiceClient:
+        case .anonymousClient, .xpcServiceClient:
             return XPCServiceClient(serviceName: endpoint.serviceName, connection: connection)
```

Comment:
Hmm, but looking at PR #23 the `XPCServiceClient`'s `createConnection()` function would result in creating a *new* anonymous connection not connecting to this one? I'm guessing this kind of works right now because the connection itself is being passed in and so `createConnection()` isn't actually getting called, but that feels far too brittle of logic.

It's likely clearer to have a distinct `XPCAnonymousClient` class to handle this case separately. Yes it's a bit more code, but it'll be easier to understand what's going on.

Also how does an anonymous client handle when its connection becomes invalid? I don't believe there's any way for it to reestablish the connection. If that's true then it's important all of the `send`/`sendMessage` functions throw a descriptive error when that happens.

Reply:
> Hmm, but looking at PR #23 the XPCServiceClient's createConnection() function would result in creating a new anonymous connection not connecting to this one

MMmm I'm not sure what you mean. When using this `forEndpoint` API, the connection is made on line 130 (from the endpoint), passed to the XPCServiceClient as a param on line 141

> I'm guessing this kind of works right now because the connection itself is being passed in and so `createConnection()` isn't actually getting called, but that feels far too brittle of logic.

That's exactly correct. Why does it feel brittle? It seems pretty reasonable to me: "If you give me my connection, I'll use it, otherwise I'll make my own connection, with blackjack and hookers]"

> It's likely clearer to have a distinct XPCAnonymousClient class to handle this case separately. Yes it's a bit more code, but it'll be easier to understand what's going on.

I've been on the fence about it. I was considering perhaps rolling all the clients together into one class, because unlike the server case, they really don't have that many differences, thus they don't really need polymorhpism that much.

> Also how does an anonymous client handle when its connection becomes invalid?

I haven't gotten that far lol. Happy path only, baby!

Reply:
> That's exactly correct. Why does it feel brittle? It seems pretty reasonable to me: "If you give me my connection, I'll use it, otherwise I'll make my own connection, with blackjack and hookers]"

Because there are other cases where a connection may need to be created, such as when it's been invalidated. And this won't work because there's no service name.

Reply:
I'm afraid I don't follow.

My desired behaviour was this: upon initialization, the client would use whatever connection was provided to it. If it wasn't provided one, it will create its own using `createConnection()`.

If the connection becomes invalidated, it'll be nil-ed out. Upon the next attempt to use  the connection, it'll cause another call to `createConnection()`.

This won't work well for anonymous connections (I need to make its implementation of `createConnection()` to throw an error), but other than that, I don't really see any issue. Could you elaborate on it?

Reply:
> I'm afraid I don't follow.

The point I was trying to make was exactly this:

> This won't work well for anonymous connections (I need to make its implementation of createConnection() to throw an error), but other than that, I don't really see any issue. Could you elaborate on it?

Reply:
How do you think we should design this?

We could have a state-machine that remembers if the anon connection was interrupted/invalidated, and all subsequent send attempts will call reply block with an error, or will call the error handler (after #17)

Similarly, can "invalidated" connection ever be reconnected? (for non-anon servers) If not, what should we do in that case? It should probably be similar to what we choose to do here.

Reply:
> Similarly, can "invalidated" connection ever be reconnected (for non-anon servers)?

The semantics of this matter a great deal. The literal answer to your exact question is "no", but I don't believe it's quite the right question to be asked.

The key to this is that `XPCClient` is not a Swift friendly wrapper around an `xpc_connection_t`. The client represents an entry point to communicate with an `XPCServer`. Under the hood any number of `xpc_connection_t` instances may over the lifetime of the client be used in order to communicate with the server. 

The way an `XPCClient` whose connection was invalidated can reestablish communication with an `XPCServer` is by simply creating another connection. In most cases what probably happened in order for the connection to have become invalidated is the process containing the server terminated and creating a new connection resulted in `launchd` relaunching the process, but that's effectively an implementation detail of XPC.

> If not, what should we do in that case? It should probably be similar to what we choose to do here.

We shouldn't do the same thing because for non-anonymous cases a new connection can be established.

> How do you think we should design the behaviour of disconnected clients (of anonymous services)?

I believe what would make sense is:
 - The initializer for `XPCAnonymousClient` always passes a non-nil `xpc_connection_t` instance in its super initializer call
     - This means the client starts with a `connection` and so calls to `getConnection()` don't result in a call to `createConnection()` unless there's been a reason the connection was nilled out, such as it being invalid
 - In `XPCServer` the function `createConnection()` now `throws`, which in turn means `getConnection()` and `sendWithReply(...)` also need to `throws`
     - In turn this means that all of the `send(...)/sendMessage(...)` functions need to `throws` which they already do so that's not an issue 
 - In `XPCAnonymousClient` the `createConnection()` implementation always throws an error, something like `connectionCannotBeReestablished`

Reply:
> Under the hood any number of xpc_connection_t instances may over the lifetime of the client be used in order to communicate with the server.

This is why I think it *is* the right question. As far as I remember, the "invalidation" error occurs when e.g. the service name is wrong, where re-connection attempts are guarenteed to fail. That what I was asking about.

> In most cases what probably happened in order for the connection to have become invalidated is the process containing the server terminated and creating a new connection resulted in launchd relaunching the process, but that's effectively an implementation detail of XPC.

Are you sure this isn't "interrupted"?

> In `XPCAnonymousClient` the `createConnection()` implementation always throws an error, something like `connectionCannotBeReestablished`

Sounds good to me. Tracking this in #37

Reply:
> Are you sure this isn't "interrupted"?

Oh damn, you're right. Just tested out this exact scenario. Sorry about that.

> This is why I think it is the right question. As far as I remember, the "invalidation" error occurs when e.g. the service name is wrong, where re-connection attempts are guarenteed to fail. That what I was asking about.

Reconnection is not guaranteed to fail on retry and I can rather easily imagine how an API user would make use of the initial failure:
1. create a client
2. try to send a message to the server, get back a `connectionInvalid` error
3. go actually install the blessed helper tool
4. try to send a message again, this should succeed

Reply:
Oh yeah, that's a good use case.